File: discomplexity/x/src/x/graphics/objects/cube.cc
#include "cube.hh"
#include "../application.hh"

namespace x
{
    namespace graphics
    {
        namespace objects
        {
            static void load(x::graphics::buffer & o)
            {
                o.target(x::graphics::buffer::array);
                o.usage(x::graphics::buffer::statical, x::graphics::buffer::draw);
            }

            static void load(x::graphics::vertices & o)
            {
                o.mode(x::graphics::vertices::triangle);

                o.push(x::graphics::vertex({-1.0f,-1.0f,-1.0f, 1.0f}, {-1.0f,  0.0f,  0.0f}));
                o.push(x::graphics::vertex({-1.0f,-1.0f, 1.0f, 1.0f}, {-1.0f,  0.0f,  0.0f}));
                o.push(x::graphics::vertex({-1.0f, 1.0f, 1.0f, 1.0f}, {-1.0f,  0.0f,  0.0f}));

                o.push(x::graphics::vertex({ 1.0f, 1.0f,-1.0f, 1.0f}, { 0.0f,  0.0f, -1.0f}));
                o.push(x::graphics::vertex({-1.0f,-1.0f,-1.0f, 1.0f}, { 0.0f,  0.0f, -1.0f}));
                o.push(x::graphics::vertex({-1.0f, 1.0f,-1.0f, 1.0f}, { 0.0f,  0.0f, -1.0f}));

                o.push(x::graphics::vertex({ 1.0f,-1.0f, 1.0f, 1.0f}, { 0.0f, -1.0f,  0.0f}));
                o.push(x::graphics::vertex({-1.0f,-1.0f,-1.0f, 1.0f}, { 0.0f, -1.0f,  0.0f}));
                o.push(x::graphics::vertex({ 1.0f,-1.0f,-1.0f, 1.0f}, { 0.0f, -1.0f,  0.0f}));

                o.push(x::graphics::vertex({ 1.0f, 1.0f,-1.0f, 1.0f}, { 0.0f,  0.0f, -1.0f}));
                o.push(x::graphics::vertex({ 1.0f,-1.0f,-1.0f, 1.0f}, { 0.0f,  0.0f, -1.0f}));
                o.push(x::graphics::vertex({-1.0f,-1.0f,-1.0f, 1.0f}, { 0.0f,  0.0f, -1.0f}));

                o.push(x::graphics::vertex({-1.0f,-1.0f,-1.0f, 1.0f}, {-1.0f,  0.0f,  0.0f}));
                o.push(x::graphics::vertex({-1.0f, 1.0f, 1.0f, 1.0f}, {-1.0f,  0.0f,  0.0f}));
                o.push(x::graphics::vertex({-1.0f, 1.0f,-1.0f, 1.0f}, {-1.0f,  0.0f,  0.0f}));

                o.push(x::graphics::vertex({ 1.0f,-1.0f, 1.0f, 1.0f}, { 0.0f, -1.0f,  0.0f}));
                o.push(x::graphics::vertex({-1.0f,-1.0f, 1.0f, 1.0f}, { 0.0f, -1.0f,  0.0f}));
                o.push(x::graphics::vertex({-1.0f,-1.0f,-1.0f, 1.0f}, { 0.0f, -1.0f,  0.0f}));

                o.push(x::graphics::vertex({-1.0f, 1.0f, 1.0f, 1.0f}, { 0.0f,  0.0f,  1.0f}));
                o.push(x::graphics::vertex({-1.0f,-1.0f, 1.0f, 1.0f}, { 0.0f,  0.0f,  1.0f}));
                o.push(x::graphics::vertex({ 1.0f,-1.0f, 1.0f, 1.0f}, { 0.0f,  0.0f,  1.0f}));

                o.push(x::graphics::vertex({ 1.0f, 1.0f, 1.0f, 1.0f}, { 1.0f,  0.0f,  0.0f}));
                o.push(x::graphics::vertex({ 1.0f,-1.0f,-1.0f, 1.0f}, { 1.0f,  0.0f,  0.0f}));
                o.push(x::graphics::vertex({ 1.0f, 1.0f,-1.0f, 1.0f}, { 1.0f,  0.0f,  0.0f}));

                o.push(x::graphics::vertex({ 1.0f,-1.0f,-1.0f, 1.0f}, { 1.0f,  0.0f,  0.0f}));
                o.push(x::graphics::vertex({ 1.0f, 1.0f, 1.0f, 1.0f}, { 1.0f,  0.0f,  0.0f}));
                o.push(x::graphics::vertex({ 1.0f,-1.0f, 1.0f, 1.0f}, { 1.0f,  0.0f,  0.0f}));

                o.push(x::graphics::vertex({ 1.0f, 1.0f, 1.0f, 1.0f}, { 0.0f,  1.0f,  0.0f}));
                o.push(x::graphics::vertex({ 1.0f, 1.0f,-1.0f, 1.0f}, { 0.0f,  1.0f,  0.0f}));
                o.push(x::graphics::vertex({-1.0f, 1.0f,-1.0f, 1.0f}, { 0.0f,  1.0f,  0.0f}));

                o.push(x::graphics::vertex({ 1.0f, 1.0f, 1.0f, 1.0f}, { 0.0f,  1.0f,  0.0f}));
                o.push(x::graphics::vertex({-1.0f, 1.0f,-1.0f, 1.0f}, { 0.0f,  1.0f,  0.0f}));
                o.push(x::graphics::vertex({-1.0f, 1.0f, 1.0f, 1.0f}, { 0.0f,  1.0f,  0.0f}));

                o.push(x::graphics::vertex({ 1.0f, 1.0f, 1.0f, 1.0f}, { 0.0f,  0.0f,  1.0f}));
                o.push(x::graphics::vertex({-1.0f, 1.0f, 1.0f, 1.0f}, { 0.0f,  0.0f,  1.0f}));
                o.push(x::graphics::vertex({ 1.0f,-1.0f, 1.0f, 1.0f}, { 0.0f,  0.0f,  1.0f}));
            }

            cube::cube(const x::string & name)
            :   x::graphics::primitive(name)
            {
                x::graphics::application & application = x::graphics::application::get();

                x::graphics::objects::load(__buffer);
                x::graphics::objects::load(__vertices);

                __program = application.program("standard");
            }

            cube::cube(x::string && name)
            :   x::graphics::primitive(x::move(name))
            {
                x::graphics::application & application = x::graphics::application::get();

                x::graphics::objects::load(__buffer);
                x::graphics::objects::load(__vertices);
                __program = application.program("standard");
            }

            cube::cube()
            :   x::graphics::primitive()
            {
                x::graphics::application & application = x::graphics::application::get();

                x::graphics::objects::load(__buffer);
                x::graphics::objects::load(__vertices);
                __program = application.program("standard");
            }

            cube::cube(const cube & o)
            :   x::graphics::primitive(o)
            {
            }

            cube::cube(cube && o) noexcept
            :   x::graphics::primitive(x::move(o))
            {
            }

            cube::~cube()
            {
            }

            cube & cube::operator=(const cube & o)
            {
                if(&o != this)
                {
                    x::graphics::primitive::operator=(o);
                }
                return *this;
            }

            cube & cube::operator=(cube && o) noexcept
            {
                if(&o != this)
                {
                    x::graphics::primitive::operator=(x::move(o));
                }
                return *this;
            }

            void cube::render(const x::graphics::matrix & model, x::uint32 level)
            {
                // x::console::out << "render cube" << x::endl;

                x::graphics::application & application = x::graphics::application::get();
                const x::graphics::camera & camera = application.camera();
                const x::graphics::light & light = application.light();
                const x::graphics::projection & projection = application.projection();
#if 1
                // __mvp =
                // x::graphics::matrix projection = projection.matrix();
                //     // x::graphics::matrix::perspective(x::math::radian(45.0f), (float32) application.width() / (float32) application.height(), 0.1f, 1000.0f);
                // x::graphics::matrix view = x::graphics::matrix::lookat(camera.position(), camera.center(), camera.up());
                // x::graphics::matrix model = x::graphics::matrix::identity();
                x::graphics::matrix mvp = model * camera.matrix() * projection.matrix();

#if 0
                glm::mat4 Projection = glm::perspective(glm::radians(45.0f), (float32) application.width() / (float32) application.height(), 0.1f, 1000.0f);

                glm::mat4 View = glm::lookAt(
                    glm::vec3(camera.position().x(),camera.position().y(),camera.position().z()), // 카메라는 (4,3,3) 에 있다. 월드 좌표에서
                    glm::vec3(0,0,0), // 그리고 카메라가 원점을 본다
                    glm::vec3(0,1,0));  // 머리가 위쪽이다 (0,-1,0 으로 해보면, 뒤집어 볼것이다)

                glm::mat4 Model = glm::mat4(1.0f);
                glm::mat4 _mvp = Projection * View * Model;

                x::console::out << _mvp << x::endl;
#endif // 0
#endif // 1

                application.bind(__vertices);
                application.bind(*__program);
                application.bind(*__program, "mvp", mvp);
                application.bind(*__program, "cameraPos", camera.position());
                application.bind(*__program, "lightPos", light.position());
                application.bind(*__program, "lightColor", light.color());
                application.bind(*__program, "ambientStrength", light.ambient());

                application.bind(__buffer);
                application.bind(0, __buffer, x::graphics::vertex::stride, 0);
                application.bind(1, __buffer, x::graphics::vertex::stride, 4 * sizeof(GLfloat));
                application.bind(2, __buffer, x::graphics::vertex::stride, 8 * sizeof(GLfloat));
                application.draw(__vertices, __indices);
                application.unbind(2);
                application.unbind(1);
                application.unbind(0);
            }
        }
    }
}

File: discomplexity/x/js/x/graphics/primitives/cube.js
import Primitive from '../primitive.js';
import Vertex from '../vertex.js';
import Vertices from '../vertices.js';
import Buffer from '../buffer.js';
import Matrix from '../matrix.js';
import Vector from '../vector.js';

export default class Cube extends Primitive
{
    static load(name, program, application)
    {
        return new Promise((resolve, reject) => {
            let o = new Cube(name, 0, 1, application.programs.get(program));
            o.serialize();
            application.objects.add(o);
            resolve(o);
        });
    }

    constructor(name, level, status, program)
    {
        super(name, level, status);

        this.mode = Vertices.triangle;
        this.target = Buffer.array;
        this.usage = [Buffer.statical, Buffer.draw];
        this.__program = program;

        this.push(new Vertex([-1.0,-1.0,-1.0, 1.0], [-1.0,  0.0,  0.0, 0.0]));
        this.push(new Vertex([-1.0,-1.0, 1.0, 1.0], [-1.0,  0.0,  0.0, 0.0]));
        this.push(new Vertex([-1.0, 1.0, 1.0, 1.0], [-1.0,  0.0,  0.0, 0.0]));

        this.push(new Vertex([ 1.0, 1.0,-1.0, 1.0], [ 0.0,  0.0, -1.0, 0.0]));
        this.push(new Vertex([-1.0,-1.0,-1.0, 1.0], [ 0.0,  0.0, -1.0, 0.0]));
        this.push(new Vertex([-1.0, 1.0,-1.0, 1.0], [ 0.0,  0.0, -1.0, 0.0]));

        this.push(new Vertex([ 1.0,-1.0, 1.0, 1.0], [ 0.0, -1.0,  0.0, 0.0]));
        this.push(new Vertex([-1.0,-1.0,-1.0, 1.0], [ 0.0, -1.0,  0.0, 0.0]));
        this.push(new Vertex([ 1.0,-1.0,-1.0, 1.0], [ 0.0, -1.0,  0.0, 0.0]));

        this.push(new Vertex([ 1.0, 1.0,-1.0, 1.0], [ 0.0,  0.0, -1.0, 0.0]));
        this.push(new Vertex([ 1.0,-1.0,-1.0, 1.0], [ 0.0,  0.0, -1.0, 0.0]));
        this.push(new Vertex([-1.0,-1.0,-1.0, 1.0], [ 0.0,  0.0, -1.0, 0.0]));

        this.push(new Vertex([-1.0,-1.0,-1.0, 1.0], [-1.0,  0.0,  0.0, 0.0]));
        this.push(new Vertex([-1.0, 1.0, 1.0, 1.0], [-1.0,  0.0,  0.0, 0.0]));
        this.push(new Vertex([-1.0, 1.0,-1.0, 1.0], [-1.0,  0.0,  0.0, 0.0]));

        this.push(new Vertex([ 1.0,-1.0, 1.0, 1.0], [ 0.0, -1.0,  0.0, 0.0]));
        this.push(new Vertex([-1.0,-1.0, 1.0, 1.0], [ 0.0, -1.0,  0.0, 0.0]));
        this.push(new Vertex([-1.0,-1.0,-1.0, 1.0], [ 0.0, -1.0,  0.0, 0.0]));

        this.push(new Vertex([-1.0, 1.0, 1.0, 1.0], [ 0.0,  0.0,  1.0, 0.0]));
        this.push(new Vertex([-1.0,-1.0, 1.0, 1.0], [ 0.0,  0.0,  1.0, 0.0]));
        this.push(new Vertex([ 1.0,-1.0, 1.0, 1.0], [ 0.0,  0.0,  1.0, 0.0]));

        this.push(new Vertex([ 1.0, 1.0, 1.0, 1.0], [ 1.0,  0.0,  0.0, 0.0]));
        this.push(new Vertex([ 1.0,-1.0,-1.0, 1.0], [ 1.0,  0.0,  0.0, 0.0]));
        this.push(new Vertex([ 1.0, 1.0,-1.0, 1.0], [ 1.0,  0.0,  0.0, 0.0]));

        this.push(new Vertex([ 1.0,-1.0,-1.0, 1.0], [ 1.0,  0.0,  0.0, 0.0]));
        this.push(new Vertex([ 1.0, 1.0, 1.0, 1.0], [ 1.0,  0.0,  0.0, 0.0]));
        this.push(new Vertex([ 1.0,-1.0, 1.0, 1.0], [ 1.0,  0.0,  0.0, 0.0]));

        this.push(new Vertex([ 1.0, 1.0, 1.0, 1.0], [ 0.0,  1.0,  0.0, 0.0]));
        this.push(new Vertex([ 1.0, 1.0,-1.0, 1.0], [ 0.0,  1.0,  0.0, 0.0]));
        this.push(new Vertex([-1.0, 1.0,-1.0, 1.0], [ 0.0,  1.0,  0.0, 0.0]));

        this.push(new Vertex([ 1.0, 1.0, 1.0, 1.0], [ 0.0,  1.0,  0.0, 0.0]));
        this.push(new Vertex([-1.0, 1.0,-1.0, 1.0], [ 0.0,  1.0,  0.0, 0.0]));
        this.push(new Vertex([-1.0, 1.0, 1.0, 1.0], [ 0.0,  1.0,  0.0, 0.0]));

        this.push(new Vertex([ 1.0, 1.0, 1.0, 1.0], [ 0.0,  0.0,  1.0, 0.0]));
        this.push(new Vertex([-1.0, 1.0, 1.0, 1.0], [ 0.0,  0.0,  1.0, 0.0]));
        this.push(new Vertex([ 1.0,-1.0, 1.0, 1.0], [ 0.0,  0.0,  1.0, 0.0]));
    }

    get name() { return this.__name; }
    get program() { return this.__program; }
    get buffer(){ return this.__buffer; }

    set mode(v){ this.__vertices.mode = v; }
    set target(v){ this.__buffer.target = v; }
    set usage(v){ this.__buffer.usage = v; }
    set program(o){ this.__program = o; }

    render(gl)
    {
        gl.useProgram(this.__program.id);
        this.__program.uniforms.forEach((value, name) => {
            if(name === "mvp")
            {
                let fovy = 45 * 0.01745329251994329576923690768489;
                let aspect = gl.canvas.width / gl.canvas.height;
                let near = 0.1;
                let far = 1000;
                let projection = Matrix.perspective(fovy, aspect, near, far);
                let view = Matrix.lookat([4, 3, 3, 0], [0, 0, 0, 0], [0, 1, 0, 0]);
                let model = Matrix.identity();
                let mvp = Matrix.multiply(Matrix.multiply(model,view), projection);
                gl.uniformMatrix4fv(value, false, new Float32Array(mvp));
            }
            else if(name === "cameraPos")
            {
                gl.uniform4fv(value, [4, 3, 3, 0]);
            }
            else if(name === "lightColor")
            {
                gl.uniform4fv(value, [1, 1, 1, 1]);
            }
            else if(name === "lightPos")
            {
                gl.uniform4fv(value, [18, 4, 3, 1]);
            }
            else if(name === "ambientStrength")
            {
                gl.uniform1f(value, 0.8);
            }
            else
            {
                throw new Error();
            }
        });
        this.__program.attributes.forEach((value, name) => {
            if(value >= 0)
            {
                gl.enableVertexAttribArray(value);
                this.buffer.bind(gl);
                if(name === "vertexPosition")
                {
                    gl.vertexAttribPointer(value, 4, gl.FLOAT, false, Vertex.stride, 0);
                }
                else if(name === "vertexNormal")
                {
                    gl.vertexAttribPointer(value, 4, gl.FLOAT, false, Vertex.stride, 4 * 4);
                }
            }
        });
        gl.drawArrays(gl.TRIANGLES, 0, this.__vertices.size);
    }
}
